### ISSUE TITLE
Index signature member typing dependent on index type

### ISSUE THREAD

**Issue by 4086606** [microsoft/TypeScript#44468]
# Suggestion

## 🔍 Search Terms

Multiple more than one index type property signature type specific member mapping TypeScript TS

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

Put further restrictions on index signatures being mapped to member types, driven by the _type of the index signature_. Basically add multiple index signatures to an interface that map different type keys to different type members. Hugely increases the power of index signatures. Blocked by #26797.

## 📃 Motivating Example

This:
```typescript
export type ColorIdentifier =
    `${string}.${string}`
    | 'background'
    | 'contrastActiveBorder' | 'contrastBorder' | 'focusBorder'
    | 'descriptionForeground' | 'errorForeground' | 'foreground';
export type ThemeScopeKey = `[${string}]`;
export interface IThemeScopedColorCustomizations {
    [colorId: ColorIdentifier]: string;
}
export interface IColorCustomizations {
    [colorIdOrThemeScope: ColorIdentifier | ThemeScopeKey]: IThemeScopedColorCustomizations | string;
}
```

Could be more accurately specified as:
```typescript
export type ColorIdentifier =
    `${string}.${string}`
    | 'background'
    | 'contrastActiveBorder' | 'contrastBorder' | 'focusBorder'
    | 'descriptionForeground' | 'errorForeground' | 'foreground';
export type ThemeScopeKey = `[${string}]`;
export interface IThemeScopedColorCustomizations {
    [colorId: ColorIdentifier]: string;
}
export interface IColorCustomizations {
    [colorId: ColorIdentifier]: string;
    [themeScope: ThemeScopeKey]: IThemeScopedColorCustomizations;
}
```

## 💻 Use Cases

The `vscode` repo is full of interface collections that could make immediate use of this..

**Comment by MartinJohns** [comment]
Sounds like a duplicate of #26797.

**Comment by 4086606** [comment]
#26797 is a blocking issue however I also don't want TypeScript to restrict object literal types (`type = {}` or `interface {}`) to one index signature (`[key: <key type>]: <member type>`).

**Comment by MartinJohns** [comment]
I'm not sure what you mean by "is a blocking issue".

> I also don't want TypeScript to restrict object literal types (type = {} or interface {}) to one index signature ([key: <key type>]: <member type>).

The linked issue does not impose a restriction like that. It's even mentioned explicitly that multiple signatures should be possible:

> since these are just index signatures, they can be added into any object type and in multiples.

**Comment by 4086606** [comment]
Correct! I've got that PR version installed staged on my VM and multiple index signatures work.
🤦‍♂️ I think I was half-asleep when I read that PR lol
